### DISCUSSION TITLE
请教如何使用docker编译出Android可以使用的wallet-core.aar文件

### DISCUSSION THREAD

**Discussion by yjgfn** [trustwallet/wallet-core#1877] [Ideas & Feature request]
ubuntu18.4
请教如何使用docker编译出Android可以使用的wallet-core.aar文件？
[How to use docker to compile the wallet-core.aar file that can be used by Android?]

多次尝试都没有成功！
[Many attempts were unsuccessful!]

我执行的命里是：./tools/android-release
cmd input：./tools/android-release

error info：
  #include <gnu/stubs.h>
           ^~~~~~~~~~~~~
  1 error generated.
  [96/483] Building CXX object CMakeFiles/TrustWalletCore.dir/src/Aion/Entry.cpp.o
  FAILED: CMakeFiles/TrustWalletCore.dir/src/Aion/Entry.cpp.o 
  ccache /usr/android-sdk-linux/ndk/21.2.6472646/toolchains/llvm/prebuilt/linux-x86_64/bin/clang++ --target=i686-none-linux-android23 --gcc-toolchain=/usr/android-sdk-linux/ndk/21.2.6472646/toolchains/llvm/prebuilt/linux-x86_64 --sysroot=/usr/android-sdk-linux/ndk/21.2.6472646/toolchains/llvm/prebuilt/linux-x86_64/sysroot  -DTrustWalletCore_EXPORTS -I../../../../../../build/local/include -I../../../../../../build/local/src/protobuf/protobuf-3.14.0/src -I../../../../../../include -I../../../../../../src -I../../../../../../jni/cpp -I../../../../../../trezor-crypto/include -isystem /usr/include -g -DANDROID -fdata-sections -ffunction-sections -funwind-tables -fstack-protector-strong -no-canonical-prefixes -mstackrealign -D_FORTIFY_SOURCE=2 -Wformat -Werror=format-security   -Wshorten-64-to-32 -O2 -DNDEBUG  -fPIC -fvisibility=hidden   -Wall -std=gnu++1z -MD -MT CMakeFiles/TrustWalletCore.dir/src/Aion/Entry.cpp.o -MF CMakeFiles/TrustWalletCore.dir/src/Aion/Entry.cpp.o.d -o CMakeFiles/TrustWalletCore.dir/src/Aion/Entry.cpp.o -c ../../../../../../src/Aion/Entry.cpp
  In file included from ../../../../../../src/Aion/Entry.cpp:7:
  In file included from ../../../../../../src/Aion/Entry.h:9:
  In file included from ../../../../../../src/Aion/../CoinEntry.h:9:
  In file included from ../../../../../../include/TrustWalletCore/TWCoinType.h:9:
  In file included from ../../../../../../include/TrustWalletCore/TWBase.h:68:
  In file included from /usr/android-sdk-linux/ndk/21.2.6472646/toolchains/llvm/prebuilt/linux-x86_64/sysroot/usr/include/c++/v1/stdbool.h:22:
  In file included from /usr/android-sdk-linux/ndk/21.2.6472646/toolchains/llvm/prebuilt/linux-x86_64/sysroot/usr/include/c++/v1/__config:215:
  /usr/include/features.h:485:10: fatal error: 'gnu/stubs.h' file not found
  #include <gnu/stubs.h>
           ^~~~~~~~~~~~~
  1 error generated.
  [97/483] Building CXX object CMakeFiles/TrustWalletCore.dir/src/Aeternity/Signer.cpp.o
  FAILED: CMakeFiles/TrustWalletCore.dir/src/Aeternity/Signer.cpp.o 
  ccache /usr/android-sdk-linux/ndk/21.2.6472646/toolchains/llvm/prebuilt/linux-x86_64/bin/clang++ --target=i686-none-linux-android23 --gcc-toolchain=/usr/android-sdk-linux/ndk/21.2.6472646/toolchains/llvm/prebuilt/linux-x86_64 --sysroot=/usr/android-sdk-linux/ndk/21.2.6472646/toolchains/llvm/prebuilt/linux-x86_64/sysroot  -DTrustWalletCore_EXPORTS -I../../../../../../build/local/include -I../../../../../../build/local/src/protobuf/protobuf-3.14.0/src -I../../../../../../include -I../../../../../../src -I../../../../../../jni/cpp -I../../../../../../trezor-crypto/include -isystem /usr/include -g -DANDROID -fdata-sections -ffunction-sections -funwind-tables -fstack-protector-strong -no-canonical-prefixes -mstackrealign -D_FORTIFY_SOURCE=2 -Wformat -Werror=format-security   -Wshorten-64-to-32 -O2 -DNDEBUG  -fPIC -fvisibility=hidden   -Wall -std=gnu++1z -MD -MT CMakeFiles/TrustWalletCore.dir/src/Aeternity/Signer.cpp.o -MF CMakeFiles/TrustWalletCore.dir/src/Aeternity/Signer.cpp.o.d -o CMakeFiles/TrustWalletCore.dir/src/Aeternity/Signer.cpp.o -c ../../../../../../src/Aeternity/Signer.cpp
  In file included from ../../../../../../src/Aeternity/Signer.cpp:7:
  In file included from ../../../../../../src/Aeternity/Signer.h:7:
  In file included from ../../../../../../src/Aeternity/Transaction.h:9:
  In file included from ../../../../../../src/Data.h:9:
  In file included from /usr/android-sdk-linux/ndk/21.2.6472646/toolchains/llvm/prebuilt/linux-x86_64/sysroot/usr/include/c++/v1/cstdint:143:
  In file included from /usr/android-sdk-linux/ndk/21.2.6472646/toolchains/llvm/prebuilt/linux-x86_64/sysroot/usr/include/c++/v1/__config:215:
  /usr/include/features.h:485:10: fatal error: 'gnu/stubs.h' file not found
  #include <gnu/stubs.h>
           ^~~~~~~~~~~~~
  1 error generated.
  [98/483] Building CXX object CMakeFiles/TrustWalletCore.dir/src/NEO/Signer.cpp.o
  FAILED: CMakeFiles/TrustWalletCore.dir/src/NEO/Signer.cpp.o 
  ccache /usr/android-sdk-linux/ndk/21.2.6472646/toolchains/llvm/prebuilt/linux-x86_64/bin/clang++ --target=i686-none-linux-android23 --gcc-toolchain=/usr/android-sdk-linux/ndk/21.2.6472646/toolchains/llvm/prebuilt/linux-x86_64 --sysroot=/usr/android-sdk-linux/ndk/21.2.6472646/toolchains/llvm/prebuilt/linux-x86_64/sysroot  -DTrustWalletCore_EXPORTS -I../../../../../../build/local/include -I../../../../../../build/local/src/protobuf/protobuf-3.14.0/src -I../../../../../../include -I../../../../../../src -I../../../../../../jni/cpp -I../../../../../../trezor-crypto/include -isystem /usr/include -g -DANDROID -fdata-sections -ffunction-sections -funwind-tables -fstack-protector-strong -no-canonical-prefixes -mstackrealign -D_FORTIFY_SOURCE=2 -Wformat -Werror=format-security   -Wshorten-64-to-32 -O2 -DNDEBUG  -fPIC -fvisibility=hidden   -Wall -std=gnu++1z -MD -MT CMakeFiles/TrustWalletCore.dir/src/NEO/Signer.cpp.o -MF CMakeFiles/TrustWalletCore.dir/src/NEO/Signer.cpp.o.d -o CMakeFiles/TrustWalletCore.dir/src/NEO/Signer.cpp.o -c ../../../../../../src/NEO/Signer.cpp
  In file included from ../../../../../../src/NEO/Signer.cpp:7:
  In file included from ../../../../../../src/NEO/Signer.h:9:
  In file included from ../../../../../../src/NEO/Address.h:9:
  In file included from /usr/android-sdk-linux/ndk/21.2.6472646/toolchains/llvm/prebuilt/linux-x86_64/sysroot/usr/include/c++/v1/string:503:
  In file included from /usr/android-sdk-linux/ndk/21.2.6472646/toolchains/llvm/prebuilt/linux-x86_64/sysroot/usr/include/c++/v1/__config:215:
  /usr/include/features.h:485:10: fatal error: 'gnu/stubs.h' file not found
  #include <gnu/stubs.h>
           ^~~~~~~~~~~~~
  1 error generated.

**Comment by hewigovens** [comment]
Please use macOS at this time, Ubuntu is not officially supported

**Reply by yjgfn** [reply]
好滴 。我再试试

**Reply by yjgfn** [reply]
> Please use macOS at this time, Ubuntu is not officially supported
我已经使用了 mac os m1编译
但安卓但始终没有成功

./tools/android-build 
~/Desktop/wallet/wallet-core/android ~/Desktop/wallet/wallet-core

> Configure project :trustwalletcore
WARNING:: This app only has 32-bit [x86] native libraries. Beginning August 1, 2019 Google Play store requires that all apps that include native libraries must provide 64-bit versions. For more information, visit https://g.co/64-bit-requirement
This app only has 32-bit [x86] native libraries. Beginning August 1, 2019 Google Play store requires that all apps that include native libraries must provide 64-bit versions. For more information, visit https://g.co/64-bit-requirement

> Task :app:lintVitalRelease
e: /Users/weiliu/.gradle/caches/modules-2/files-2.1/org.jetbrains.kotlin/kotlin-stdlib-jdk7/1.6.10/e1c380673654a089c4f0c9f83d0ddfdc1efdb498/kotlin-stdlib-jdk7-1.6.10.jar!/META-INF/kotlin-stdlib-jdk7.kotlin_module: Module was compiled with an incompatible version of Kotlin. The binary version of its metadata is 1.6.0, expected version is 1.4.0.
e: /Users/weiliu/.gradle/caches/modules-2/files-2.1/org.jetbrains.kotlin/kotlin-stdlib/1.6.10/b8af3fe6f1ca88526914929add63cf5e7c5049af/kotlin-stdlib-1.6.10.jar!/META-INF/kotlin-stdlib.kotlin_module: Module was compiled with an incompatible version of Kotlin. The binary version of its metadata is 1.6.0, expected version is 1.4.0.
e: /Users/weiliu/.gradle/caches/modules-2/files-2.1/org.jetbrains.kotlin/kotlin-stdlib-common/1.6.10/c118700e3a33c8a0d9adc920e9dec0831171925/kotlin-stdlib-common-1.6.10.jar!/META-INF/kotlin-stdlib-common.kotlin_module: Module was compiled with an incompatible version of Kotlin. The binary version of its metadata is 1.6.0, expected version is 1.4.0.

> Task :trustwalletcore:generateJsonModelRelease FAILED
release|armeabi-v7a :CMake Warning at /Users/weiliu/Library/Android/sdk/ndk/23.1.7779620/build/cmake/android-legacy.toolchain.cmake:416 (message):
release|armeabi-v7a :  An old version of CMake is being used that cannot automatically detect
release|armeabi-v7a :  compiler attributes.  Compiler identification is being bypassed.  Some
release|armeabi-v7a :  values may be wrong or missing.  Update to CMake 3.19 or newer to use
release|armeabi-v7a :  CMake's built-in compiler identification.
release|armeabi-v7a :Call Stack (most recent call first):
release|armeabi-v7a :  /Users/weiliu/Library/Android/sdk/ndk/23.1.7779620/build/cmake/android.toolchain.cmake:55 (include)
release|armeabi-v7a :  /Users/weiliu/Library/Android/sdk/cmake/3.18.1/share/cmake-3.18/Modules/CMakeDetermineSystem.cmake:93 (include)
release|armeabi-v7a :  CMakeLists.txt:3 (project)
release|armeabi-v7a :CMake Warning at /Users/weiliu/Library/Android/sdk/ndk/23.1.7779620/build/cmake/android-legacy.toolchain.cmake:416 (message):
release|armeabi-v7a :  An old version of CMake is being used that cannot automatically detect
release|armeabi-v7a :  compiler attributes.  Compiler identification is being bypassed.  Some
release|armeabi-v7a :  values may be wrong or missing.  Update to CMake 3.19 or newer to use
release|armeabi-v7a :  CMake's built-in compiler identification.
release|armeabi-v7a :Call Stack (most recent call first):
release|armeabi-v7a :  /Users/weiliu/Library/Android/sdk/ndk/23.1.7779620/build/cmake/android.toolchain.cmake:55 (include)
release|armeabi-v7a :  /Users/weiliu/Desktop/wallet/wallet-core/android/trustwalletcore/.cxx/cmake/release/armeabi-v7a/CMakeFiles/3.18.1-g262b901/CMakeSystem.cmake:6 (include)
release|armeabi-v7a :  /Users/weiliu/Desktop/wallet/wallet-core/android/trustwalletcore/.cxx/cmake/release/armeabi-v7a/CMakeFiles/CMakeTmp/CMakeLists.txt:2 (project)
release|armeabi-v7a :CMake Warning at /Users/weiliu/Library/Android/sdk/ndk/23.1.7779620/build/cmake/android-legacy.toolchain.cmake:416 (message):
release|armeabi-v7a :  An old version of CMake is being used that cannot automatically detect
release|armeabi-v7a :  compiler attributes.  Compiler identification is being bypassed.  Some
release|armeabi-v7a :  values may be wrong or missing.  Update to CMake 3.19 or newer to use
release|armeabi-v7a :  CMake's built-in compiler identification.
release|armeabi-v7a :Call Stack (most recent call first):
release|armeabi-v7a :  /Users/weiliu/Library/Android/sdk/ndk/23.1.7779620/build/cmake/android.toolchain.cmake:55 (include)
release|armeabi-v7a :  /Users/weiliu/Desktop/wallet/wallet-core/android/trustwalletcore/.cxx/cmake/release/armeabi-v7a/CMakeFiles/3.18.1-g262b901/CMakeSystem.cmake:6 (include)
release|armeabi-v7a :  /Users/weiliu/Desktop/wallet/wallet-core/android/trustwalletcore/.cxx/cmake/release/armeabi-v7a/CMakeFiles/CMakeTmp/CMakeLists.txt:2 (project)
release|armeabi-v7a :CMake Error at /Users/weiliu/Library/Android/sdk/cmake/3.18.1/share/cmake-3.18/Modules/FindPackageHandleStandardArgs.cmake:165 (message):
release|armeabi-v7a :  Could NOT find Boost (missing: Boost_INCLUDE_DIR)
release|armeabi-v7a :Call Stack (most recent call first):
release|armeabi-v7a :  /Users/weiliu/Library/Android/sdk/cmake/3.18.1/share/cmake-3.18/Modules/FindPackageHandleStandardArgs.cmake:458 (_FPHSA_FAILURE_MESSAGE)
release|armeabi-v7a :  /Users/weiliu/Library/Android/sdk/cmake/3.18.1/share/cmake-3.18/Modules/FindBoost.cmake:2177 (find_package_handle_standard_args)
release|armeabi-v7a :  CMakeLists.txt:44 (find_package)
release|armeabi-v7a :  CMakeLists.txt:50 (find_host_package)

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':trustwalletcore:generateJsonModelRelease'.
> error when building with cmake using /Users/weiliu/Desktop/wallet/wallet-core/CMakeLists.txt: Build command failed.
  Error while executing process /Users/weiliu/Library/Android/sdk/cmake/3.18.1/bin/cmake with arguments {-H/Users/weiliu/Desktop/wallet/wallet-core -DCMAKE_FIND_ROOT_PATH=/Users/weiliu/Desktop/wallet/wallet-core/android/trustwalletcore/.cxx/cmake/release/prefab/armeabi-v7a/prefab -DCMAKE_TOOLCHAIN_FILE=/Users/weiliu/Library/Android/sdk/ndk/23.1.7779620/build/cmake/android.toolchain.cmake -DANDROID_ABI=armeabi-v7a -DANDROID_NDK=/Users/weiliu/Library/Android/sdk/ndk/23.1.7779620 -DANDROID_PLATFORM=android-23 -DCMAKE_ANDROID_ARCH_ABI=armeabi-v7a -DCMAKE_ANDROID_NDK=/Users/weiliu/Library/Android/sdk/ndk/23.1.7779620 -DCMAKE_EXPORT_COMPILE_COMMANDS=ON -DCMAKE_LIBRARY_OUTPUT_DIRECTORY=/Users/weiliu/Desktop/wallet/wallet-core/android/trustwalletcore/build/intermediates/cmake/release/obj/armeabi-v7a -DCMAKE_RUNTIME_OUTPUT_DIRECTORY=/Users/weiliu/Desktop/wallet/wallet-core/android/trustwalletcore/build/intermediates/cmake/release/obj/armeabi-v7a -DCMAKE_MAKE_PROGRAM=/Users/weiliu/Library/Android/sdk/cmake/3.18.1/bin/ninja -DCMAKE_SYSTEM_NAME=Android -DCMAKE_SYSTEM_VERSION=23 -DCMAKE_BUILD_TYPE=Release -B/Users/weiliu/Desktop/wallet/wallet-core/android/trustwalletcore/.cxx/cmake/release/armeabi-v7a -GNinja}
  -- Detecting C compiler ABI info
  -- Detecting C compiler ABI info - done
  -- Check for working C compiler: /Users/weiliu/Library/Android/sdk/ndk/23.1.7779620/toolchains/llvm/prebuilt/darwin-x86_64/bin/clang - skipped
  -- Detecting C compile features
  -- Detecting C compile features - done
  -- Detecting CXX compiler ABI info
  -- Detecting CXX compiler ABI info - done
  -- Check for working CXX compiler: /Users/weiliu/Library/Android/sdk/ndk/23.1.7779620/toolchains/llvm/prebuilt/darwin-x86_64/bin/clang++ - skipped
  -- Detecting CXX compile features
  -- Detecting CXX compile features - done
  -- Configuring incomplete, errors occurred!
  See also "/Users/weiliu/Desktop/wallet/wallet-core/android/trustwalletcore/.cxx/cmake/release/armeabi-v7a/CMakeFiles/CMakeOutput.log".
  
  CMake Warning at /Users/weiliu/Library/Android/sdk/ndk/23.1.7779620/build/cmake/android-legacy.toolchain.cmake:416 (message):
    An old version of CMake is being used that cannot automatically detect
    compiler attributes.  Compiler identification is being bypassed.  Some
    values may be wrong or missing.  Update to CMake 3.19 or newer to use
    CMake's built-in compiler identification.
  Call Stack (most recent call first):
    /Users/weiliu/Library/Android/sdk/ndk/23.1.7779620/build/cmake/android.toolchain.cmake:55 (include)
    /Users/weiliu/Library/Android/sdk/cmake/3.18.1/share/cmake-3.18/Modules/CMakeDetermineSystem.cmake:93 (include)
    CMakeLists.txt:3 (project)
  
  
  CMake Warning at /Users/weiliu/Library/Android/sdk/ndk/23.1.7779620/build/cmake/android-legacy.toolchain.cmake:416 (message):
    An old version of CMake is being used that cannot automatically detect
    compiler attributes.  Compiler identification is being bypassed.  Some
    values may be wrong or missing.  Update to CMake 3.19 or newer to use
    CMake's built-in compiler identification.
  Call Stack (most recent call first):
    /Users/weiliu/Library/Android/sdk/ndk/23.1.7779620/build/cmake/android.toolchain.cmake:55 (include)
    /Users/weiliu/Desktop/wallet/wallet-core/android/trustwalletcore/.cxx/cmake/release/armeabi-v7a/CMakeFiles/3.18.1-g262b901/CMakeSystem.cmake:6 (include)
    /Users/weiliu/Desktop/wallet/wallet-core/android/trustwalletcore/.cxx/cmake/release/armeabi-v7a/CMakeFiles/CMakeTmp/CMakeLists.txt:2 (project)
  
  
  CMake Warning at /Users/weiliu/Library/Android/sdk/ndk/23.1.7779620/build/cmake/android-legacy.toolchain.cmake:416 (message):
    An old version of CMake is being used that cannot automatically detect
    compiler attributes.  Compiler identification is being bypassed.  Some
    values may be wrong or missing.  Update to CMake 3.19 or newer to use
    CMake's built-in compiler identification.
  Call Stack (most recent call first):
    /Users/weiliu/Library/Android/sdk/ndk/23.1.7779620/build/cmake/android.toolchain.cmake:55 (include)
    /Users/weiliu/Desktop/wallet/wallet-core/android/trustwalletcore/.cxx/cmake/release/armeabi-v7a/CMakeFiles/3.18.1-g262b901/CMakeSystem.cmake:6 (include)
    /Users/weiliu/Desktop/wallet/wallet-core/android/trustwalletcore/.cxx/cmake/release/armeabi-v7a/CMakeFiles/CMakeTmp/CMakeLists.txt:2 (project)
  
  
  CMake Error at /Users/weiliu/Library/Android/sdk/cmake/3.18.1/share/cmake-3.18/Modules/FindPackageHandleStandardArgs.cmake:165 (message):
    Could NOT find Boost (missing: Boost_INCLUDE_DIR)
  Call Stack (most recent call first):
    /Users/weiliu/Library/Android/sdk/cmake/3.18.1/share/cmake-3.18/Modules/FindPackageHandleStandardArgs.cmake:458 (_FPHSA_FAILURE_MESSAGE)
    /Users/weiliu/Library/Android/sdk/cmake/3.18.1/share/cmake-3.18/Modules/FindBoost.cmake:2177 (find_package_handle_standard_args)
    CMakeLists.txt:44 (find_package)
    CMakeLists.txt:50 (find_host_package)
  
  


* Try:
Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output. Run with --scan to get full insights.

* Get more help at https://help.gradle.org

BUILD FAILED in 1m 3s
32 actionable tasks: 7 executed, 24 from cache, 1 up-to-date

**Reply by hewigovens** [reply]
See the error, you need to install `boost`

**Reply by yjgfn** [reply]
已安装boost了
